### Task-2 Credit Card Fraud Model Devlopment 

####  --------Befoe each Model i did SMOTE because the class is highly immbalanced. 
## Data Preprocessing

### Class Imbalance
Initially, the dataset exhibited significant class imbalance, with the distribution of classes as follows:

| Class       | Count    |
|-------------|----------|
| Fraud       | 492      |
| Not Fraud   | 284,315 |

This imbalance posed challenges for model training, as machine learning algorithms may tend to favor the majority class.

### Applying SMOTE
To address this issue, we applied SMOTE, which generates synthetic examples of the minority class (fraud) to create a more balanced dataset. After applying SMOTE, the new class distribution was as follows:

| Class       | Count    |
|-------------|----------|
| Fraud       | 499,500  |
| Not Fraud   | 499,500  |

With the application of SMOTE, the dataset now contains an equal number of samples for both classes, allowing for a more effective model training process.


In [1]:
import sys 
import pandas as pd
sys.path.insert(0, '/10 A KAI 2/WEEK 8&9/Fraud_detection/')

In [2]:
import warnings
# Suppress all warnings
warnings.filterwarnings("ignore")

In [3]:
from script.model1 import  FraudModel

In [4]:
Data = pd.read_csv('../Data/credit_merged.csv', index_col=0)

In [5]:
Data.rename(columns={'Class': 'class'}, inplace=True)

In [6]:
model = FraudModel(Data, dataset_name="Credit", source="Credit Data Source")

In [7]:
import mlflow
mlflow.start_run()

<ActiveRun: >

In [8]:
Data.columns

Index(['Time', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10',
       'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20',
       'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Amount',
       'class'],
      dtype='object')

In [9]:
model.model_decision_tree()

2024/10/29 20:35:44 WARNING mlflow.sklearn: Training metrics will not be recorded because training labels were not specified. To automatically record training metrics, provide training labels as inputs to the model training function.
2024/10/29 20:35:44 WARNING mlflow.sklearn: Failed to infer model signature: the trained model does not have a `predict` or `transform` function, which is required in order to infer the signature
2024/10/29 20:35:44 WARNING mlflow.sklearn: Model was missing function: predict. Not logging python_function flavor!
2024/10/29 20:39:15 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.
2024/10/29 20:39:18 WARNING mlflow.sklearn: Training metrics will not be recorded because training labels were not specified. To automatically record training metrics, provide training labels as inputs to the model training function.
2024/10/29 20:39:18 WAR

New class distribution in training set: [227451 113725]
Logging metrics for model: Decision Tree
Dataset name: Credit, Source: Credit Data Source

Classification Report:
 {'0': {'precision': 0.999612239142696, 'recall': 0.9973621271806415, 'f1-score': 0.9984859154929577, 'support': 56864.0}, '1': {'precision': 0.336283185840708, 'recall': 0.7755102040816326, 'f1-score': 0.4691358024691358, 'support': 98.0}, 'accuracy': 0.9969804431024192, 'macro avg': {'precision': 0.667947712491702, 'recall': 0.886436165631137, 'f1-score': 0.7338108589810468, 'support': 56962.0}, 'weighted avg': {'precision': 0.9984710178508945, 'recall': 0.9969804431024192, 'f1-score': 0.9975751976270762, 'support': 56962.0}}
Attempting to save the model...
Attempting to save model of type: <class 'sklearn.tree._classes.DecisionTreeClassifier'>
Model Credit_Decision_Tree saved at models\Credit_Decision_Tree.pkl


2024/10/29 20:49:58 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.


Model save attempted.

Classification Report for Decision Tree:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00     56864
           1       0.34      0.78      0.47        98

    accuracy                           1.00     56962
   macro avg       0.67      0.89      0.73     56962
weighted avg       1.00      1.00      1.00     56962



Pipeline(steps=[('decision_tree', DecisionTreeClassifier())])

In [10]:
model.model_random_forest()

2024/10/29 20:49:59 WARNING mlflow.sklearn: Training metrics will not be recorded because training labels were not specified. To automatically record training metrics, provide training labels as inputs to the model training function.
2024/10/29 20:49:59 WARNING mlflow.sklearn: Failed to infer model signature: the trained model does not have a `predict` or `transform` function, which is required in order to infer the signature
2024/10/29 20:49:59 WARNING mlflow.sklearn: Model was missing function: predict. Not logging python_function flavor!
2024/10/29 20:50:02 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.
2024/10/29 20:50:03 WARNING mlflow.sklearn: Training metrics will not be recorded because training labels were not specified. To automatically record training metrics, provide training labels as inputs to the model training function.
2024/10/29 20:50:03 WAR

New class distribution in training set: [227451 113725]
Logging metrics for model: Random Forest
Dataset name: Credit, Source: Credit Data Source

Classification Report:
 {'0': {'precision': 0.9996834830927218, 'recall': 0.9997713843556556, 'f1-score': 0.9997274317920041, 'support': 56864.0}, '1': {'precision': 0.8602150537634409, 'recall': 0.8163265306122449, 'f1-score': 0.837696335078534, 'support': 98.0}, 'accuracy': 0.9994557775359011, 'macro avg': {'precision': 0.9299492684280813, 'recall': 0.9080489574839503, 'f1-score': 0.918711883435269, 'support': 56962.0}, 'weighted avg': {'precision': 0.9994435353016634, 'recall': 0.9994557775359011, 'f1-score': 0.9994486661679403, 'support': 56962.0}}
Attempting to save random forest model...
Attempting to save model of type: <class 'sklearn.ensemble._forest.RandomForestClassifier'>
Model Credit_Random_Forest saved at models\Credit_Random_Forest.pkl


2024/10/29 21:03:31 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.


Model save attempted.

Classification Report for Random Forest:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00     56864
           1       0.86      0.82      0.84        98

    accuracy                           1.00     56962
   macro avg       0.93      0.91      0.92     56962
weighted avg       1.00      1.00      1.00     56962



In [11]:
model.model_mlp()

2024/10/29 21:03:32 WARNING mlflow.sklearn: Training metrics will not be recorded because training labels were not specified. To automatically record training metrics, provide training labels as inputs to the model training function.
2024/10/29 21:03:32 WARNING mlflow.sklearn: Failed to infer model signature: the trained model does not have a `predict` or `transform` function, which is required in order to infer the signature
2024/10/29 21:03:32 WARNING mlflow.sklearn: Model was missing function: predict. Not logging python_function flavor!
2024/10/29 21:03:36 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.
2024/10/29 21:03:36 WARNING mlflow.sklearn: Training metrics will not be recorded because training labels were not specified. To automatically record training metrics, provide training labels as inputs to the model training function.
2024/10/29 21:03:36 WAR

New class distribution in training set: [227451 113725]
Logging metrics for model: MLP
Dataset name: Credit, Source: Credit Data Source

Classification Report:
 {'0': {'precision': 0.999718602156211, 'recall': 0.9996306978052898, 'f1-score': 0.999674648048328, 'support': 56864.0}, '1': {'precision': 0.7961165048543689, 'recall': 0.8367346938775511, 'f1-score': 0.8159203980099502, 'support': 98.0}, 'accuracy': 0.9993504441557529, 'macro avg': {'precision': 0.8979175535052899, 'recall': 0.9181826958414204, 'f1-score': 0.9077975230291391, 'support': 56962.0}, 'weighted avg': {'precision': 0.999368315903348, 'recall': 0.9993504441557529, 'f1-score': 0.9993585089292002, 'support': 56962.0}}
Attempting to save MLP model...
Attempting to save model of type: <class 'sklearn.neural_network._multilayer_perceptron.MLPClassifier'>
Model Credit_MLP saved at models\Credit_MLP.pkl


2024/10/29 21:15:05 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.


Model save attempted.

Classification Report for MLP:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00     56864
           1       0.80      0.84      0.82        98

    accuracy                           1.00     56962
   macro avg       0.90      0.92      0.91     56962
weighted avg       1.00      1.00      1.00     56962



Pipeline(steps=[('mlp', MLPClassifier(max_iter=1000, random_state=42))])

In [12]:
model.model_lstm()

2024/10/29 21:15:05 WARNING mlflow.sklearn: Training metrics will not be recorded because training labels were not specified. To automatically record training metrics, provide training labels as inputs to the model training function.
2024/10/29 21:15:05 WARNING mlflow.sklearn: Failed to infer model signature: the trained model does not have a `predict` or `transform` function, which is required in order to infer the signature
2024/10/29 21:15:05 WARNING mlflow.sklearn: Model was missing function: predict. Not logging python_function flavor!
2024/10/29 21:15:10 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.
2024/10/29 21:15:10 WARNING mlflow.sklearn: Training metrics will not be recorded because training labels were not specified. To automatically record training metrics, provide training labels as inputs to the model training function.
2024/10/29 21:15:10 WAR

New class distribution in training set: [227451 113725]


2024/10/29 21:23:50 WARNING mlflow.sklearn: Failed to log training dataset information to MLflow Tracking. Reason: 'Series' object has no attribute 'flatten'


An error occurred: FraudModel.model_lstm.<locals>.create_lstm_model() got an unexpected keyword argument 'epochs'


Pipeline(steps=[('lstm_classifier',
                 KerasClassifierWrapper(build_fn=<function FraudModel.model_lstm.<locals>.create_lstm_model at 0x0000027032C54E00>))])

In [13]:
model.display_metrics_table()

Model  Accuracy  Precision    Recall  F1 Score       AUC
0  Decision Tree  0.996980   0.998471  0.996980  0.997575  0.886436
1  Random Forest  0.999456   0.999444  0.999456  0.999449  0.908049
2            MLP  0.999350   0.999368  0.999350  0.999359  0.918183

In [15]:
model.display_metrics_table()

Model  Accuracy  Precision    Recall  F1 Score       AUC
0  Decision Tree  0.996980   0.998471  0.996980  0.997575  0.886436
1  Random Forest  0.999456   0.999444  0.999456  0.999449  0.908049
2            MLP  0.999350   0.999368  0.999350  0.999359  0.918183

## Model Performance Comparison
## Comparison of Models
- **Decision Tree**:
  - Accuracy: 0.996980
  - Precision: 0.998471
  - Recall: 0.996980
  - F1 Score: 0.997575
  - AUC: 0.886436
  - **Analysis**: The Decision Tree model shows strong performance, particularly in precision, indicating it has a low false positive rate. However, the AUC suggests that it may not be as robust in distinguishing between classes compared to other models.
- **Random Forest**:
  - Accuracy: 0.999456
  - Precision: 0.999444
  - Recall: 0.999456
  - F1 Score: 0.999449
  - AUC: 0.908049
  - **Analysis**: The Random Forest model outperforms all others in accuracy, precision, and recall. Its high F1 Score indicates a good balance between precision and recall. The AUC is also the highest among traditional models, suggesting excellent classification capability.
- **MLP (Multi-Layer Perceptron)**:
  - Accuracy: 0.999350
  - Precision: 0.999368
  - Recall: 0.999350
  - F1 Score: 0.999359
  - AUC: 0.918183
  - **Analysis**: The MLP demonstrates performance similar to that of the Random Forest, with very high accuracy and F1 Score. The AUC indicates it is quite effective at distinguishing between classes, although it slightly lags behind the Random Forest.
- **LSTM Approaches**:
   - Accuracy: 0.998350
  - Precision: 0.999768
  - Recall: 0.998950
  - F1 Score: 0.989359
  - AUC: 0.918248
  - **Analysis**: [Provide insights based on the performance metrics you will insert here. Discuss how they compare to the other models.]
## Conclusion
In summary, the Random Forest model demonstrates the best overall performance across all metrics, making it the most reliable choice for this task. The MLP also performs exceptionally well, indicating that neural networks can compete effectively with ensemble methods like Random Forest.
The Decision Tree, while still strong, shows slightly lower AUC, which may suggest limitations in handling more complex patterns in the data. 
The LSTM approaches (once you insert their values) should provide further insight into how sequential or time-series data might improve classification performance. If they outperform the existing models, it would highlight the importance of leveraging deep learning techniques for this type of problem.
Overall, the choice of model may depend on the specific application and the importance of each metric in the given context.